Relax scale to prevent saturation in conv/linear. Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25667 Relax scale to prevent saturation in conv/linear. Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in**

Relax scale and zero-point for activations to ensure that fbgemm implementations of conv and linear do not saturate due to 16 bit intermediate accumulation.

Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in
handling layouts for quantized ops in addition to saturation/quantization errors.

Differential Revision: [D17140498](https://our.internmc.facebook.com/intern/diff/D17140498/)

raghuramank10000 committed:
Relax scale to prevent saturation in conv/linear. Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in

Relax scale and zero-point for activations to ensure that fbgemm implementations of conv and linear do not saturate due to 16 bit intermediate accumulation.

Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in
handling layouts for quantized ops in addition to saturation/quantization errors.

Differential Revision: [D17140498](https://our.internmc.facebook.com/intern/diff/D17140498/)

dskhudia commented on test/test_quantized_models.py:
```diff
@@ -0,0 +1,26 @@
+        torch.quantization.convert(qModel)
+        out_q = qModel(eval_data)
+        SQNR = 20 * torch.log10(torch.norm(out_ref) / torch.norm(out_ref - out_q))
+        self.assertGreater(SQNR, 30, msg='Quantized model numerics diverge from float, expect SQNR > 30 dB')
```
How did you decide on 30?

raghuramank100 replied:
I want a value that is as high as possible. With 8 bits of precision the maximum we could get is around 40 dB (~5 dB per bit). With a cascade of many layers, what is achievable is below this. I see 35-40 dB when I run it for different random seeds. The goal is to have a tight check so that we can find any numerics errors.

dskhudia replied:
tight check might result in flaky test depending on what random values are generated.

raghuramank100 replied:
There is no randomness in the test. I fix the random seed so that the numerics do not change.

dskhudia commented on torch/quantization/observer.py:
```diff
@@ -97,9 +105,43 @@ class MinMaxObserver(ObserverBase):
+        #   FBGEMM uses vpmaddubsw instruction to multiply activations (uint8_t) and
+        #   weights (int8_t).
+        #
+        #  https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_maddubs_epi16&expand=3284,3530
```
May be for this part ... we can just point to the comment in the qconv.cpp file rather than repeating it.

raghuramank100 replied:
Ok, will abbreviate it and point to the comment

llyfacebook commented on torch/quantization/observer.py:
```diff
@@ -126,6 +168,8 @@ def observer(observer_cls, **kwargs):
 
 def default_observer(**kwargs):
+    # Restrict activations to be in the range (0,127)
+    kwargs.setdefault("reduce_range", True)
```
Should it be False by default?

raghuramank100 replied:
I think it should be true so that when someone uses the default observer, they get the best possible accuracy. Setting reduce_range to be true does that.

llyfacebook replied:
Hmm, I just thought reducing range is not the normal case, since it reduce the number of levels.

raghuramank100 replied:
We need the default to work without any changes for x86 quantization and provide the best performance. Leaving this as an option to the user runs the risk of having poorer accuracy for quantized models.

hx89 commented on test/test_quantization.py:
```diff
@@ -636,9 +636,13 @@ def checkQuantized(model):
+           qscheme=st.sampled_from((torch.per_tensor_affine, torch.per_tensor_symmetric)),
+           reduce_range=st.booleans())
+    def test_minmax_observer(self, qdtype, qscheme, reduce_range):
+        # reduce_range cannot be true for symmetric quantization with uint8
```
Just wonder why reduce_range cannot be true for symmetric quantization with uint8? Seems we can use qmin=, qmax=127 with zero_point = 64 to do the quantization.

raghuramank100 replied:
For symmetric quantization, we are constrained to use zero-point = 128 (middle of the 8 bit range). With that value, we cannot eliminate overflow as we will have values greater than 128, which can cause overflow when multiplied with values close to 255.

dskhudia replied:
Are we reducing 1-bit of precision from both weights and activations?

raghuramank100 replied:
No, only for the activations. The weight_observer is unchanged and uses all 8 bits

raghuramank10000 committed:
Update on "Relax scale to prevent saturation in conv/linear. Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in"


Relax scale and zero-point for activations to ensure that fbgemm implementations of conv and linear do not saturate due to 16 bit intermediate accumulation.

Add test to verify precision of numerics of quantized model with updated observer. This test catches errors in
handling layouts for quantized ops in addition to saturation/quantization errors.

Differential Revision: [D17140498](https://our.internmc.facebook.com/intern/diff/D17140498/)

jianyuh commented on test/test_quantized_models.py:
```diff
@@ -0,0 +1,26 @@
+        torch.quantization.convert(qModel)
+        out_q = qModel(eval_data)
+        SQNRdB = 20 * torch.log10(torch.norm(out_ref) / torch.norm(out_ref - out_q))
+        self.assertGreater(SQNRdB, 30, msg='Quantized model numerics diverge from float, expect SQNR > 30 dB')
```
Maybe add a comment to explain the choice of 30 dB here.

raghuramank100 replied:
Will do prior to landing

jianyuh commented on torch/quantization/observer.py:
```diff
@@ -126,6 +143,8 @@ def observer(observer_cls, **kwargs):
     return partial(observer_cls, **kwargs)
 
 def default_observer(**kwargs):
+    # Restrict activations to be in the range (0,127)
```
Wondering the reason for restricting the reduced precision on the activation instead of the weight? How will it affect the accuracy result if we instead restrict the range of weight (int8) to be (-64, 63)?

dskhudia replied:
From the overflow point of view, we have to reduce one bit for either one of them. I think the results would be similar if we reduce 1-bit for weights.

raghuramank100 replied:
I agree, I kept the weights at 8 bits because with QAT we can learn weights at a higher precision compared to 7. For post training quant, both choices lead to similar results.

jianyuh reviewed (approved):
Please add the comment for the choice of 30 dB in the unit test. Other than that, the PR looks good to me.
